### DISCUSSION TITLE
AWS Xray exporter?

### DISCUSSION THREAD

**Discussion by swgoldwood** [open-telemetry/opentelemetry-cpp#1259] [General]
Has anybody exported their traces to AWS Xray, or know of a way of accomplishing this, such as via some intermediate process? Would really appreciate any guidance!

Many Thanks

**Comment by esigo** [comment]
You may export your traces to  otel-collector and then  the collector export to other backends including [AWS Xray](https://aws-otel.github.io/docs/getting-started/x-ray#configuring-the-otlp-receiver).

**Reply by swgoldwood** [reply]
Thank you! This is the way

**Reply by esigo** [reply]
Also you can find an example using the otel-collector [here](https://github.com/open-telemetry/opentelemetry-cpp/tree/main/examples/otlp).

**Comment by lalitb** [comment]
Moving to discussion.

**Reply by alfianabdi** [reply]
How to generate AWS XRay trace id ? 
https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/aws/src/main/java/io/opentelemetry/sdk/extension/aws/trace/AwsXrayIdGenerator.java

My goal is to include trace from nginx (https://github.com/open-telemetry/opentelemetry-cpp-contrib/tree/main/instrumentation/nginx) but it seems that the trace id is not accepted by xray. I keep getting this error from exporter

```
2022-06-02T08:52:27.820Z        debug   awsxrayexporter@v0.47.0/awsxray.go:99   Error translating span. {"kind": "exporter", "name": "awsxray", "error": "invalid xray traceid: a1f9a39e43cdb1fcf28a1c063516b4fa"}
```

**Comment by noliva** [comment]
So once there's an implementation of the AWS Xray ID generator, we should be able to send Otel traces to an XRay backend, but we need the custom Id generator for the reasons @alfianabdi mentioned.

**Reply by alfianabdi** [reply]
I tried to create AwsXrayIdGenerator here. 
https://github.com/alfianabdi/opentelemetry-cpp/blob/main/sdk/src/trace/aws_xray_id_generator.cc

It works with my nginx opentelemetry instrumentation testing.
However I am not sure if it follow good code standard. I am also not sure where should I put the function, I believe it should not be inside opentelemetry-cpp repository.

**Comment by dluongiop** [comment]
Is there any update on when opentelemetry-cpp will be able to generate xray compatible traceid for exporting traces to aws xray? 

It seems to me that if the awsxray expoerter would generate the xray trace id and incorporate the original otlp trace id into the xray trace document would be a better solution.

**Reply by lalitb** [reply]
opentelemetry-cpp provides the ability to plugin a custom TraceId generator. You can use that to generate the x-ray compatible TraceId. Refer : https://opentelemetry-cpp.readthedocs.io/en/latest/otel_docs/classopentelemetry_1_1sdk_1_1trace_1_1IdGenerator.html?highlight=idgenerator

**Reply by dluongiop** [reply]
Thank you.  This worked very well.
